build/proc-ctl.py

```diff
@@ -0,0 +1,271 @@
+
+    stop = grp.add_parser('stop', help='Stop a running subprocess')
+    stop.add_argument('--ctl-dir',
+                      help='The control directory for the subprocess',
+                      required=True,
+                      type=Path)
+    stop.add_argument('--stop-wait',
+                      help='Number of seconds to wait for stopping',
+                      type=float,
+                      default=5)
+    stop.add_argument('--if-not-running',
+                      help='Action to take if the child is not running',
+                      choices=['fail', 'ignore'],
+                      default='fail')
+
+    ll_run = grp.add_parser('__run')
+    ll_run.add_argument('--ctl-dir', type=Path, required=True)
+    ll_run.add_argument('child_command', nargs=argparse.REMAINDER)
+
+    return parser
+
+
+if TYPE_CHECKING:
+    StartCommandArgs = NamedTuple('StartCommandArgs', [
+        ('command', Literal['start']),
+        ('ctl_dir', Path),
+        ('cwd', Path),
+        ('child_command', Sequence[str]),
+        ('spawn_wait', int),
+    ])
+
+    StopCommandArgs = NamedTuple('StopCommandArgs', [
+        ('command', Literal['stop']),
+        ('ctl_dir', Path),
+        ('stop_wait', float),
+        ('if_not_running', Literal['fail', 'ignore']),
+    ])
+
+    _RunCommandArgs = NamedTuple('_RunCommandArgs', [
+        ('command', Literal['__run']),
+        ('child_command', Sequence[str]),
+        ('ctl_dir', Path),
+    ])
+
+    CommandArgs = Union[StartCommandArgs, StopCommandArgs, _RunCommandArgs]
+
+    _ResultType = TypedDict('_ResultType', {
+        'exit': 'str | int | None',
+        'error': 'str | None'
+    })
+
+
+def parse_argv(argv: 'Sequence[str]') -> 'CommandArgs':
+    parser = create_parser()
+    args = parser.parse_args(argv)
+    return cast('CommandArgs', args)
+
+
+class _ChildControl:
+
+    def __init__(self, ctl_dir: Path) -> None:
+        self._ctl_dir = ctl_dir
+
+    @property
+    def pid_file(self):
+        """The file containing the child PID"""
+        return self._ctl_dir / 'pid.txt'
+
+    @property
+    def result_file(self):
+        """The file containing the exit result"""
+        return self._ctl_dir / 'exit.json'
+
+    def set_pid(self, pid: int):
+        self.pid_file.write_text(str(pid))
+
+    def get_pid(self) -> 'int | None':
+        try:
+            txt = self.pid_file.read_text()
+        except FileNotFoundError:
+            return None
+        return int(txt)
+
+    def set_exit(self, exit: 'str | int | None', error: 'str | None') -> None:
+        self.result_file.write_text(json.dumps({'exit': exit, 'error': error}))
+        self._unlink(self.pid_file)
+
+    def get_result(self) -> 'None | _ResultType':
+        try:
+            txt = self.result_file.read_text()
+        except FileNotFoundError:
+            return None
+        try:
+            return json.loads(txt)
+        except json.JSONDecodeError:
+            return self.get_result()
```

Comment:
This is to catch partial writes to the result file. The recursive call is just a loop to "retry" the read on the results file. In the worst case, this can hit a recursion limit, but that's unlikely.